Linter and formatting scripts

j-hui opened the pull request:
I added some convenience wrappers to the linter (hlint) and formatter (brittany) we're using.

I will write more documentation in the README later, but essentially this will allow us to run `git lint` and `git fmt` to lint and format our code. It supports (1) passing explicit file names to format/lint, and (2) passing a commit hash/branch name, to lint/format all files since then. By default, it just lints/formats all the code modified since the last commit.

hmontero1205 commented:
Do you want to go through the existing codebase right now and run these? There is existing code that isn't formatted. It may be better for the blame to be attributed to this PR than for it to be scattered..

j-hui commented:
I'd like for this to be merged in first. I have another PR in the works
where I string together the whole compiler pipeline. I was thinking of
doing all this formatting then.

On Sat, Oct 9, 2021, 1:04 PM Hans Montero ***@***.***> wrote:

> Do you want to go through the existing codebase right now and run these?
> There is existing code that isn't formatted. It may be better for the blame
> to be attributed to this PR than for it to be scattered..
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ssm-lang/sslang/pull/11#issuecomment-939328401>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2A5DFGXCMMQ6MDWQZ6J43UGBYZFANCNFSM5FUZUWAQ>
> .
>

j-hui commented on README.md:
```diff
@@ -1,46 +1,137 @@
+
+### Setup Development Toolchain
+
+You can install this compiler's build toolchain using [GHCup][ghcup], Haskell's toolchain manager, with the following command:
```
Discussed in 1-to-1 meeting; in summary: GHCup is a toolchain manager (manages installation of GHC, Stack, Cabal, and HLS), while Stack is a project + package manager (manages installation of package dependencies for each package). GHCup seems to be the recommended way to install these tools [according to Haskell.org](https://www.haskell.org/downloads/); GHCup provides some more rationale in their FAQ: https://www.haskell.org/ghcup/about/#faq

j-hui replied:
(I'll add a note about this + pointers to those links in the README)

j-hui commented on README.md:
```diff
@@ -1,46 +1,137 @@
+
+This will run a short but interactive script; make sure to let it know where it should add the `PATH` variable. It will also ask if you would like to install [HLS][hls] (Haskell Language Server); you should do so if you would like to set up IDE features with an LSP-compatible text editor.
+
+The GHCup setup script may also detect that dependencies are missing, and ask you to install them; make sure to do so before proceeding. For instance, if you are running Ubuntu 20.10:
```
Stack's install script is secretly doing this: https://github.com/commercialhaskell/stack/blob/master/etc/scripts/get-stack.sh#L168

GHCup ultimately puts the onus on the user to install system dependencies (though it provides helpful suggestions) because it doesn't want to be in the business of scripting around each system's package manager.

j-hui commented on README.md:
```diff
@@ -1,46 +1,137 @@
+Convenience scripts are provided under the [`scripts`](./scripts/) subdirectory, to help lint, format, and build this respoitory's code. As long as your current working directory is within this repo, you may invoke these scripts directly,
+or you may set up Git aliases for easier access.
+
+To setup git aliases:
```
This should work from any directory; it's putting the `../.gitconfig` path into `.git/config` (which is why there is the `../`). But I'll make a note of it.

j-hui commented on README.md:
`````diff
@@ -1,46 +1,137 @@
-make
-````
+```
+git watch
`````
No; will clarify

j-hui commented:
@sedwards-lab addressed your feedback, but the README is looking rather lengthy. I'm wondering if we can move parts of it to the `CONTRIBUTING.md` document? Or should I create a table of contents?

sedwards-lab reviewed (approved):
Quite a lot of restructuring. I like the longer, more thoughtful CONTRIBUTING file. Thanks!
